docs/how-to/optimize.md

```diff
@@ -0,0 +1,150 @@
+                system_prompt_content=self.config.system_prompt_content,
+                question=data["question"],
+            )
+        )
+        answer = await llm.generate(prompt=input_prompt)
+        return RandomQuestionPipelineResult(answer=answer)
+```
+
+### Define the Data Loader
+
+Next, we define the data loader. We'll use Ragbits generation stack to create an artificial data loader:
+
+
+```python
+from ragbits.evaluate.loaders.base import DataLoader, DataT
+from ragbits.core.llms.litellm import LiteLLM
+from ragbits.core.prompt import Prompt
+from pydantic import BaseModel
+from omegaconf import OmegaConf
+
+
+class DatasetGenerationPromptInput(BaseModel):
+    topic: str
+
+
+class DatasetGenerationPrompt(Prompt[DatasetGenerationPromptInput]):
+    system_prompt = "Be a provider of random questions on a topic specified by the user."
+    user_prompt = "Generate a question about {{ topic }}"
+
+
+class RandomQuestionsDataLoader(DataLoader):
+    async def load(self) -> DataT:
+        questions = []
+        llm = LiteLLM()
+        for _ in range(self.config.num_questions):
+            question = await llm.generate(
+                DatasetGenerationPrompt(DatasetGenerationPromptInput(topic=self.config.question_topic))
+            )
+            questions.append({"question": question})
+        return questions
+
+
+dataloader_config = OmegaConf.create(
+    {"num_questions": 10, "question_topic": "conspiracy theories"}
+)
+dataloader = RandomQuestionsDataLoader(dataloader_config)
+```
+
+### Define the Metrics and Run the Experiment
+
+```python
+from typing import Any
+import tiktoken
+from ragbits.evaluate.optimizer import Optimizer
+from ragbits.evaluate.metrics.base import Metric, MetricSet, ResultT
+from omegaconf import OmegaConf
+
+
+class TokenCountMetric(Metric):
+    def compute(self, results: list[ResultT]) -> dict[str, Any]:
+        encoding = tiktoken.get_encoding("cl100k_base")
+        num_tokens = [len(encoding.encode(out.answer)) for out in results]
+        return {"num_tokens": sum(num_tokens) / len(num_tokens)}
+
+
+metrics = MetricSet(TokenCountMetric())
+
+optimization_cfg = OmegaConf.create(
+    {"direction": "minimize", "n_trials": 4}
+)
+optimizer = Optimizer(optimization_cfg)
+
+optimized_params = OmegaConf.create(
+    {
+        "system_prompt_content": {
+            "optimize": True,
+            "choices": [
+                "Be a friendly bot answering user questions. Be as concise as possible",
+                "Be a silly bot answering user questions. Use as few tokens as possible",
+                "Be informative and straight to the point",
+                "Respond to user questions in as few words as possible",
+            ],
+        }
+    }
+)
+configs_with_scores = optimizer.optimize(
+    pipeline_class=RandomQuestionRespondPipeline,
+    config_with_params=optimized_params,
+    metrics=metrics,
+    dataloader=dataloader,
+)
+print(configs_with_scores)
+```
```

Comment:
At the end of this how-to we should show results and briefly describe what we achieved

Reply:
in that case - probably we would need to ensure reproducability of experiments - so that user won't see different output than presented in howto - will do it